### PR TITLE
Add optional displayName parameter for kubernetes clusters

### PR DESCRIPTION
**Pull request by sydorovdmytro**
This change adds a new optional parameter `displayName`. If `displayName` is defined, it will be used instead of the cluster name. This might be useful when we have multiple clusters with the same name.

Signed-off-by: Dmytro Sydorov <dmytro.sydorov@bonial.com>

### CODE REVIEW

File: .changeset/nine-glasses-obey.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+'@backstage/plugin-kubernetes': patch
+'@backstage/plugin-kubernetes-backend': patch
+'@backstage/plugin-kubernetes-common': patch
+---
+
+Add optional displayName parameter for kubernetes clusters
```
Comment (pjungermann) [inline .changeset/nine-glasses-obey.md]: best to put displayName into ticks

File: docs/features/kubernetes/configuration.md [hunk 2]
```diff
@@ -94,7 +95,12 @@ The base URL to the Kubernetes control plane. Can be found by using the
 ##### `clusters.\*.name`
 
 A name to represent this cluster, this must be unique within the `clusters`
-array. Users will see this value in the Software Catalog Kubernetes plugin.
+array. Users will see this value in the Software Catalog Kubernetes plugin if `clusters.\*.displayName` is not set.
```
Comment (pjungermann) [inline docs/features/kubernetes/configuration.md]: you might need a different line break here. Might fail validation 🤔 
I would probably put each sentence in a separate line

File: plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.test.ts [hunk 3]
```diff
@@ -49,6 +49,7 @@ describe('ConfigClusterLocator', () => {
     expect(result).toStrictEqual([
       {
         name: 'cluster1',
+        displayName: undefined,
```
Comment (pjungermann) [inline plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.test.ts:52]: these should be unnecessary usually

File: plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.ts [hunk 4]
```diff
@@ -50,6 +51,9 @@ export class ConfigClusterLocator implements KubernetesClustersSupplier {
         if (c.has('dashboardParameters')) {
           clusterDetails.dashboardParameters = c.get('dashboardParameters');
         }
+        if (c.has('displayName')) {
+          clusterDetails.displayName = c.get('displayName');
+        }
```
Comment (pjungermann) [inline plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.ts:56]: why is this needed? You already set it in line 41. You can remove 54-56.

line 51-53 could be simplified, too

File: plugins/kubernetes-backend/src/service/KubernetesFanOutHandler.ts [hunk 5]
```diff
@@ -336,6 +336,9 @@ export class KubernetesFanOutHandler {
     if (clusterDetails.dashboardParameters) {
       objects.cluster.dashboardParameters = clusterDetails.dashboardParameters;
     }
+    if (clusterDetails.displayName) {
+      objects.cluster.displayName = clusterDetails.displayName;
```
Comment (pjungermann) [inline plugins/kubernetes-backend/src/service/KubernetesFanOutHandler.ts:340]: if is not really necessary here. You could simplify this and just assign the value.

File: plugins/kubernetes/src/api/types.ts [hunk 6]
```diff
@@ -33,6 +33,7 @@ export interface KubernetesApi {
   getClusters(): Promise<
     {
       name: string;
+      displayName?: string | undefined;
```
Comment (pjungermann) [inline plugins/kubernetes/src/api/types.ts]: `displayName?: string` is enough. The `?` already means the same.

Basically:
`displayName?: string` == `displayName: string | undefined`

File: plugins/kubernetes/src/components/Cluster/Cluster.tsx [hunk 7]
```diff
@@ -46,13 +46,15 @@ import { PodNamesWithMetricsContext } from '../../hooks/PodNamesWithMetrics';
 
 type ClusterSummaryProps = {
   clusterName: string;
+  displayName?: string | undefined;
```
Comment (pjungermann) [inline plugins/kubernetes/src/components/Cluster/Cluster.tsx]: same here

File: plugins/kubernetes/src/components/Cluster/Cluster.tsx [hunk 8]
```diff
@@ -73,7 +75,9 @@ const ClusterSummary = ({
         spacing={0}
       >
         <Grid item xs>
-          <Typography variant="h3">{clusterName}</Typography>
+          <Typography variant="h3">
+            {displayName ? displayName : clusterName}
```
Comment (pjungermann) [inline plugins/kubernetes/src/components/Cluster/Cluster.tsx]: `displayName ?? clusterName`

File: plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.ts (hunk 4 above)
Comment (sydorovdmytro) [reply plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.ts:56]: @pjungermann if I skip this condition, the `displayName: undefined` will be passed to every `clusterDetails` object. So we will get something like
```
{
    name: 'other-cluster',
    displayName: undefined,
    authProvider: 'google',
}
```

I would keep it as is. The same applied for KubernetesFanOutHandler

File: .changeset/nine-glasses-obey.md (hunk 1 above)
Comment (sydorovdmytro) [reply .changeset/nine-glasses-obey.md]: done

File: docs/features/kubernetes/configuration.md (hunk 2 above)
Comment (sydorovdmytro) [reply docs/features/kubernetes/configuration.md]: done

File: plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.test.ts (hunk 3 above)
Comment (sydorovdmytro) [reply plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.test.ts:52]: If we don't set it here:
```
 FAIL   @backstage/plugin-kubernetes-backend  plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.test.ts
  ● ConfigClusterLocator › one clusters returns one cluster details

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 1

    @@ -1,9 +1,10 @@
      Array [
        Object {
          "authProvider": "serviceAccount",
          "caData": undefined,
    +     "displayName": undefined,
    ```

File: plugins/kubernetes/src/api/types.ts (hunk 6 above)
Comment (sydorovdmytro) [reply plugins/kubernetes/src/api/types.ts]: done

File: plugins/kubernetes/src/components/Cluster/Cluster.tsx (hunk 7 above)
Comment (sydorovdmytro) [reply plugins/kubernetes/src/components/Cluster/Cluster.tsx]: done

File: plugins/kubernetes/src/components/Cluster/Cluster.tsx (hunk 8 above)
Comment (sydorovdmytro) [reply plugins/kubernetes/src/components/Cluster/Cluster.tsx]: done

File: plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.ts (hunk 4 above)
Comment (pjungermann) [reply plugins/kubernetes-backend/src/cluster-locator/ConfigClusterLocator.ts:56]: hmm.. it is anyways a property which defaults to `undefined` but ok 🤔
